src/Eventso.Subscription.Kafka.DeadLetter.Postgres/PoisonEventStore.cs

```diff
@@ -0,0 +1,386 @@
+            var offsets = new long[topicPartitionOffsets.Count];
+            foreach (var (index, topicPartitionOffset) in Enumerate(topicPartitionOffsets))
+            {
+                topics[index] = topicPartitionOffset.Topic;
+                partitions[index] = topicPartitionOffset.Partition.Value;
+                offsets[index] = topicPartitionOffset.Offset.Value;
+            }
+
+            await using var connection = _connectionFactory.ReadWrite();
+
+            await using var command = new NpgsqlCommand(
+                @"
+DELETE FROM eventso_dlq.poison_events pe
+USING UNNEST (@topics, @partitions, @offsets) AS input(topic, partition, ""offset"")
+WHERE pe.topic = input.topic AND pe.partition = input.partition AND pe.""offset"" = input.""offset"";",
+                connection)
+            {
+                Parameters =
+                {
+                    new NpgsqlParameter<string[]>("topics", topics),
+                    new NpgsqlParameter<int[]>("partitions", partitions),
+                    new NpgsqlParameter<long[]>("offsets", offsets),
+                }
+            };
+
+            await connection.OpenAsync(token);
+
+            await command.ExecuteNonQueryAsync(token);
+        }
+
+        public async IAsyncEnumerable<StoredPoisonEvent> GetEventsForRetrying(
+            string topic,
+            [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            await using var connection = _connectionFactory.ReadOnly();
+
+            await using var command = new NpgsqlCommand(
+                @"
+
+SELECT
+    partition,
+    ""offset"",
+    pe_heads.key,
+    value,
+    creation_timestamp,
+    header_keys,
+    header_values,
+    last_failure_timestamp,
+    last_failure_reason,
+    total_failure_count
+FROM eventso_dlq.poison_events pe_heads
+INNER JOIN (
+    SELECT key, MIN(""offset"") AS min_offset
+    FROM eventso_dlq.poison_events
+    WHERE topic = @topic
+    GROUP BY key
+) heads
+ON pe_heads.key = heads.key and pe_heads.""offset"" = heads.min_offset
+WHERE
+    pe_heads.total_failure_count <= @maxAllowedFailureCount
+    AND NOW() - pe_heads.last_failure_timestamp > @minIntervalBetweenRetries;",
+                connection)
+            {
+                Parameters =
+                {
+                    new NpgsqlParameter<string>("topic", topic),
+                    new NpgsqlParameter<int>("maxAllowedFailureCount", _maxAllowedFailureCount),
+                    new NpgsqlParameter<TimeSpan>("minIntervalBetweenRetries", _minIntervalBetweenRetries)
+                }
+            };
+
+            await connection.OpenAsync(cancellationToken);
+
+            var reader = await command.ExecuteReaderAsync(cancellationToken);
+            while (await reader.ReadAsync(cancellationToken))
+            {
+                var headerKeys = reader.GetFieldValue<string[]>(5);
+                var headerValues = reader.GetFieldValue<byte[][]>(6);
```

Comment:
its more effective in some cases to read values from db in the same order as requested, also reading values here is optional and depends on keys count

Reply:
done it